File: app/src/main/java/com/mikifus/padland/SaferWebView/SaferWebViewClient.java
package com.mikifus.padland.SaferWebView;

import android.net.Uri;
import android.text.TextUtils;
import android.util.Log;
import android.webkit.WebResourceResponse;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.io.InputStream;
import java.io.StringBufferInputStream;

/**
   * Implements whitelisting on host name
   */
public class SaferWebViewClient extends WebViewClient {

    protected String[] hostsWhitelist;

    public SaferWebViewClient(String[] hostsWhitelist){
        super();
        this.hostsWhitelist = hostsWhitelist;
    }

    @Override
    public WebResourceResponse shouldInterceptRequest(final WebView view, String url) {
        boolean isvalid = isValidHost(url);
        if (isvalid) {
            return super.shouldInterceptRequest(view, url);
        } else {
            return getWebResourceResponseFromString();
        }
    }

    protected WebResourceResponse getWebResourceResponseFromString() {
        Log.w("SaferWebViewClient", "Blocked a JS request to an external domains.");
        return getUtf8EncodedWebResourceResponse(new StringBufferInputStream("alert('!NO!')"));
    }

    protected WebResourceResponse getUtf8EncodedWebResourceResponse(InputStream data) {
        return new WebResourceResponse("text/css", "UTF-8", data);
    }


    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        return isValidHost(url);
    }

    protected boolean isValidHost(String url){
        if (!TextUtils.isEmpty(url)) {
            final String host = Uri.parse(url).getHost();
            for (String whitelistedHost: hostsWhitelist){
                if (whitelistedHost.equalsIgnoreCase(host)){
                    return true;
                }
            }
        }
        return false;
    }
}

File: app/src/main/java/com/mikifus/padland/SaferWebView/SaferWebViewClient.kt
package com.mikifus.padland.SaferWebView

import android.net.Uri
import android.os.Build
import android.text.TextUtils
import android.util.Log
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.annotation.RequiresApi
import java.io.ByteArrayInputStream

/**
 * Implements whitelisting on host name
 */
open class SaferWebViewClient(protected var hostsWhitelist: Array<String?>) : WebViewClient() {
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun shouldInterceptRequest(view: WebView, request: WebResourceRequest): WebResourceResponse? {
        val isValid = isValidHost(request.url.toString())
        return if (isValid) {
            super.shouldInterceptRequest(view, request)
        } else {
            webResourceResponseFromString
        }
    }

    override fun shouldInterceptRequest(view: WebView, url: String): WebResourceResponse? {
        val isValid = isValidHost(url)
        return if (isValid) {
            super.shouldInterceptRequest(view, url)
        } else {
            webResourceResponseFromString
        }
    }

    protected open val webResourceResponseFromString: WebResourceResponse?
        get() {
            Log.w("SaferWebViewClient", "Blocked a JS request to an external domains.")
            return getUtf8EncodedWebResourceResponse(ByteArrayInputStream("alert('!NO!')".toByteArray()))
        }

    protected fun getUtf8EncodedWebResourceResponse(data: ByteArrayInputStream?): WebResourceResponse {
        return WebResourceResponse("text/css", "UTF-8", data)
    }

    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        return isValidHost(url)
    }

    protected open fun isValidHost(url: String?): Boolean {
        if (!TextUtils.isEmpty(url)) {
            val host = Uri.parse(url).host
            for (whitelistedHost in hostsWhitelist!!) {
                if (whitelistedHost.equals(host, ignoreCase = true)) {
                    return true
                }
            }
        }
        return false
    }
}